Add shims for `gettid`-esque functions

tgross35 opened the pull request:
Various platforms provide a function to return the current OS thread ID, but they all use a slightly different name. Add shims for these functions for Apple, FreeBSD, and Windows, with tests to account for a few more platforms that are not yet supported by Miri.

These should be useful in general but will also help support printing the OS thread ID in panic messages [1].

[1]: https://github.com/rust-lang/rust/pull/115746

tgross35 commented on src/shims/env.rs:
```diff
@@ -110,8 +110,24 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        self.get_tid(self.eval_context_ref().machine.threads.active_thread())
+    }
+
+    /// Get an "OS" thread ID for any thread.
+    fn get_tid(&self, thread: ThreadId) -> u32 {
+        let this = self.eval_context_ref();
+        let index = thread.to_u32();
+
+        // todo: is this also true on Windows?
+        // Compute a TID for this thread, ensuring that the main thread has PID == TID.
+        this.get_pid().strict_add(index)
+    }
```
@ChrisDenton is there any relationship between the PID and TIDs on Windows?

This (preexisting) logic works for Unix because the first TID for a process is the same as the PID, but if that logic isn't true on Windows then I'll need to update this. If it is true on Windows, then I think the `CreateThread` shim needs to get updated to return the same TID logic in `lpThreadId` (hence the failing test).

ChrisDenton replied:
On Windows thread IDs and process IDs are unrelated.

RalfJung replied:
But it doesn't hurt to use the Unix scheme on Windows as well I guess?

tgross35 replied:
Oops, I replied before seeing your message Ralf. Apparently it seems like a Linux thing more specifically than a Unix thing, I can't find documentation for other platforms about a relationship between PID and TID.

I just updated to only offset by PID on Linux, but do you prefer I put it back on all platforms and update the address saved by `CreateThread` instead?

tgross35 replied:
Just checked a handful of different platforms on cfarm:

Linux

```text
main tid 2212321 0x21c1e1
main pid 2212321 0x21c1e1
```

FreeBSD: 

```text
main tid 0 0
main pid 23750 0x5cc6
```

OpenBSD:

```text
main tid 137368 0x21898
main pid 93069 0x16b8d
```

NetBSD:

```text
main tid 5057 0x13c1
main pid 5057 0x13c1
```

MacOS:

```text
main tid 7763176 0x7674e8
main pid 51507 0xc933
```

Based on that nonscientific experiment I'll add NetBSD to the list of platforms that says tid is an offset from pid. For the others it's probably actually better that they don't line up, so users don't rely on that non-cross-platform behavior.

tgross35 commented:
Updated to include the `SYS_gettid` syscall

tgross35 commented:
Rebased because of the test failures but that didn't help; I don't know why this seems to have an effect on `alloc-access-tracking.rs`.

RalfJung commented on src/shims/unix/linux_like/syscall.rs:
```diff
@@ -53,6 +55,11 @@ pub fn syscall<'tcx>(
             ecx.write_int(result.to_i32()?, dest)?;
         }
+        num if num == sys_gettid => {
+            let [] = check_exact_vararg_count("syscall(SYS_gettid, ...)", varargs)?;
```
Please remain consistent with the other variadic functions -- I think last time my conclusion was that we should never error when there are extra arguments, for anything variadic, so something like `check_exact_vararg_count` deliberately does not exist.

RalfJung commented on src/shims/unix/env.rs:
```diff
@@ -275,15 +276,56 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-        this.assert_target_os("linux", "gettid");
+        let target_os = &self.eval_context_ref().tcx.sess.target.os;
+
+        // Various platforms have a similar function with different names.
+        match (target_os.as_ref(), link_name.as_str()) {
+            ("linux" | "android", "gettid") => (),
+            ("openbsd", "getthrid") => (),
+            ("freebsd", "pthread_getthreadid_np") => (),
+            ("netbsd", "_lwp_self") => (),
+            (_, name) => panic!("`{name}` is not supported on {target_os}"),
+        };
+
```
I don't think it's worth having such a table, this should be handled by the dispatch logic before already. In here we should just assert that this is truly a unix platform.

RalfJung commented on src/shims/env.rs:
```diff
@@ -110,8 +110,29 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+
+    /// Get an "OS" thread ID for the current thread.
+    fn get_current_tid(&self) -> u32 {
+        self.get_tid(self.eval_context_ref().machine.threads.active_thread())
```
Please follow the usual pattern.
```suggestion
        let this = self.eval_context_ref();
        this.get_tid(this.machine.threads.active_thread())
```

RalfJung commented on src/shims/unix/env.rs:
```diff
@@ -275,15 +276,56 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    ) -> InterpResult<'tcx, Scalar> {
+        let this = self.eval_context_mut();
+
+        let target_vendor = &this.tcx.sess.target.vendor;
+        assert_eq!(
+            target_vendor, "apple",
```
We usually query the OS, not the vendor... so please remain consistent and use `assert_target_os`, there's no point in having different logic for this in different places.

tgross35 replied:
I had this because it does seem to be supported on iOS and other Apple targets, but I suppose they probably aren't otherwise supported anyway :)

RalfJung replied:
Yeah the only apple target we support atm is macOS.

RalfJung commented on src/shims/unix/env.rs:
```diff
@@ -275,15 +276,56 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
-        let index = this.machine.threads.active_thread().to_u32();
+    /// The Apple-specific `int pthread_threadid_np(pthread_t thread, uint64_t *thread_id)`, which
+    /// allows querying the ID for arbitrary threads.
```
```suggestion
    /// allows querying the ID for arbitrary threads identified by their pthread_t.
```

RalfJung commented on src/shims/unix/env.rs:
```diff
@@ -275,15 +276,56 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+            // Null thread ID indicates that we are querying the active thread.
+            this.machine.threads.active_thread()
+        } else {
+            let Ok(thread) = this.thread_id_try_from(thread) else {
```
```suggestion
            // Our pthread_t is just the raw ThreadId.
            let Ok(thread) = this.thread_id_try_from(thread) else {
```

RalfJung commented on src/shims/unix/env.rs:
```diff
@@ -275,15 +276,56 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-        // Compute a TID for this thread, ensuring that the main thread has PID == TID.
-        let tid = this.get_pid().strict_add(index);
+        let tid_dest = this.deref_pointer_as(tid_op, this.machine.layouts.u64)?;
+        this.write_int(this.get_tid(thread), &tid_dest)?;
```
Please let-bind the `this.get_tid(thread)`, so that the core logic is a bit easier to spot among the boilerplate.

RalfJung commented on src/shims/unix/linux_like/syscall.rs:
```diff
@@ -53,6 +55,11 @@ pub fn syscall<'tcx>(
         }
+        num if num == sys_gettid => {
+            let [] = check_exact_vararg_count("syscall(SYS_gettid, ...)", varargs)?;
+            let result = ecx.unix_gettid(Symbol::intern("gettid"))?;
```
```suggestion
            let result = ecx.unix_gettid(Symbol::intern("SYS_gettid"))?;
```

RalfJung reviewed (commented):
Lots of little comments :)
@rustbot author

tgross35 commented on tests/pass-dep/shims/gettid.rs:
```diff
@@ -0,0 +1,167 @@
+    from_libc
+}
+
+/// Specific platforms can query the tid of arbitrary threads; test that here.
+#[cfg(any(target_vendor = "apple", windows))]
+mod queried {
```
Windows and apple have a way to get the thread ID of threads other than the current one, which is being tested here. But to do this you need the `pthread_t` or the `HANDLE`, which as far as I know we can't get from `std`.

I'll comment this a bit better

tgross35 commented on tests/pass-dep/shims/gettid.rs:
```diff
@@ -0,0 +1,167 @@
+//@ ignore-target: illumos solaris wasm
```
I originally thought they didn't have a function equivalent to `gettid` (returning OS TID), but on the r-l/r thread the maintainers confirmed that the value returned by `pthread_self` is the OS TID on those platforms. So I'll just expand the tests.

tgross35 commented on src/shims/extern_static.rs:
```diff
@@ -66,7 +66,7 @@ impl<'tcx> MiriMachine<'tcx> {
                     &["__cxa_thread_atexit_impl", "__clock_gettime64"],
                 )?;
-                Self::weak_symbol_extern_statics(ecx, &["getrandom", "statx"])?;
+                Self::weak_symbol_extern_statics(ecx, &["getrandom", "gettid", "statx"])?;
```
It seems I needed to increment by 2, not sure if that is expected

RalfJung replied:
Ah right, there's two new allocations: one to hold the static `gettid`, and one that represent the function which that static points to.

tgross35 commented:
I believe I addressed all of the concerns. I also updated the Windows and Mac implementations (and tests) since they return gracefully rather than crashing with null or (on Windows) invalid pointers.

@rustbot ready

tgross35 commented on src/shims/unix/env.rs:
```diff
@@ -310,6 +310,12 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
 
+        let tid_dest = this.read_pointer(tid_op)?;
+        if this.ptr_is_null(tid_dest)? {
+            // If NULL is passed, an error is immediately returned
```
The possible error codes are documented https://www.manpagez.com/man/3/pthread_threadid_np/, the priority (null check before valid thread check) is observed

RalfJung replied:
That's apparently a BSD manpage, not an Apple one... but it's better than no reference. Would be good to have some sort of link somewhere in the function for what we used as reference.

tgross35 replied:
It is from MacOS (see the table at the very bottom). Source for reference https://github.com/apple-oss-distributions/libpthread/blob/1ebf56b3a702df53213c2996e5e128a535d2577e/src/pthread.c#L1118-L1146 (there is also a manpage in that repo)

RalfJung replied:
> It is from MacOS (see the table at the very bottom). 

Ah, I missed that. Please link to that then :)

tgross35 commented on tests/pass-dep/shims/gettid.rs:
```diff
@@ -0,0 +1,187 @@
+        // Ensure that we got a different thread ID.
+        assert_ne!(spawned_tid_from_handle, current_tid);
+
+        // Ensure that we get the same result from `gettid` and from querying a thread's handle
+        assert_ne!(spawned_tid_from_handle, 0);
+        assert_eq!(spawned_tid_from_handle, spawned_tid_from_self);
```
In Windows it indicates the error condition, but it's already checked in the relevant function. It still shouldn't happen on Unix but the check probably isn't very useful there, so I've remove this.

tgross35 commented on tests/pass-dep/shims/gettid.rs:
```diff
@@ -133,7 +133,8 @@ mod queried {
             assert_ne!(res, WAIT_FAILED, "failed to join thread");
 
-            invalid_handle_tid = GetThreadId(usize::MAX as HANDLE);
+            // Ensure we return 0 for invalid threads rather than crashing
```
Is this Windows-only or is something similar done for invalid `pthread_t`s on other platforms?

RalfJung replied:
This is specifically for windows handles.

tgross35 commented:
Switched back to the old Windows behavior and the other small things, on to the next round

@rustbot review

RalfJung commented:
This looks great, thanks! I have pushed a commit with a small comment tweak. Please squash the commits using `./miri squash`. Then write `@rustbot ready` after you force-pushed the squashed PR.

@rustbot author

tgross35 commented:
Thanks!

RalfJung commented:
We just did a rustc-push sync, so ordinarily it would take around a week until this one gets synced. Let me know if something is blocked on this PR then I can do an earlier sync.

tgross35 commented:
https://github.com/rust-lang/rust/pull/115746 still has ~6 days left on fcp so there isn't really any rush here, but thank you for offering
